### PR TITLE
Fix index of TabBarView when decrementing

### PR DESCRIPTION
**Pull request by HeavenOSK**
When the `TabBarView` is scrolled, the value of `pageController.page` with `floor` is passed to `tabController`. But when `TabBarView` is scrolled to the left, the value of `pageController.page` is greatly reduced due to the use of `floor`.
 
Therefore, I fixed to use `round` instead of `floor`, following the behavior for left swipe.
 
before | after
:-: | :-:
<video src='https://user-images.githubusercontent.com/19836917/130793500-15d42eb7-b1d6-44d4-9773-3813740a541d.mov' width=180/> | <video src='https://user-images.githubusercontent.com/19836917/130803738-f90a4040-c9df-4df5-90d1-b6767e0389e6.mov' width=180/>

## Related issue
- https://github.com/flutter/flutter/issues/88875
 
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by justinmc** [comment]
@HeavenOSK Thanks for the PR and the videos in the description.  This change looks good, but you'll need a test.

There are some existing tests that do almost what you want, so it shouldn't be too hard to write one.  See [this test](https://github.com/flutter/flutter/blob/1d862ec7eaaa3e34a162c53b5d504c11d10543fa/packages/flutter/test/material/tabs_test.dart#L497) in tabs_test.dart.  Maybe if you reduce the distance in the moveTos you can reproduce the bug?  As long as it fails on master and passes here with your change.

Let me know if that seems to work.

**Comment by HeavenOSK** [comment]
@justinmc 
Thanks for your support. I added a test based on the link you shared with me. 
https://github.com/flutter/flutter/pull/88878/commits/7620e5d33e2589391c2888cea750f08415573ccf
 
Please review it when you have time 🙏 
 
Thanks

File: packages/flutter/lib/src/material/tabs.dart
Comment (nt4f04uNd) [inline packages/flutter/lib/src/material/tabs.dart:1342]: It seems to change the behavior of the offset, does it?

Comment (HeavenOSK) [reply packages/flutter/lib/src/material/tabs.dart:1342]: Yes it does. To be precise, I adapted the behavior to the end of flinging. 
 
When flinging ends, `round` is used.
https://github.com/flutter/flutter/blob/7620e5d33e2589391c2888cea750f08415573ccf/packages/flutter/lib/src/material/tabs.dart#L1344

File: packages/flutter/test/material/tabs_test.dart
Comment (justinmc) [inline packages/flutter/test/material/tabs_test.dart:641]:
```diff
- expect(tabController.index == indexOfB, true);
+ expect(tabController.index, indexOfB);
```

This will give more useful logging information if it fails.

File: packages/flutter/test/material/tabs_test.dart
Comment (justinmc) [inline packages/flutter/test/material/tabs_test.dart:644]: Something is not right either with this test or my understanding.  I would have thought that the `600` here could have been less than 200, the value used in the [test that I gave as an example](https://github.com/flutter/flutter/blob/1d862ec7eaaa3e34a162c53b5d504c11d10543fa/packages/flutter/test/material/tabs_test.dart#L519), because this PR allows the detection of smaller page flings.  Is that right?

Also, if I put a log in your tabController listener, I see that it is called twice.  I'm not sure why that is, maybe because the fling is so big?

**Review by justinmc — COMMENTED** [review]
I'm struggling to understand the test, see my comment below.

File: packages/flutter/test/material/tabs_test.dart
Comment (HeavenOSK) [reply packages/flutter/test/material/tabs_test.dart]: Thanks for your advice. I fixed it at https://github.com/flutter/flutter/pull/88878/commits/435709655c1acb19b77e51b9030892b932259bbd .

File: packages/flutter/test/material/tabs_test.dart
Comment (HeavenOSK) [reply packages/flutter/test/material/tabs_test.dart:644]: @justinmc 
Sorry late response 🙏 
 
 > I see that it is called twice.
 
I think this behavior is as normal. When I made the video for this PR, the listener was called twice for a fling.
 
> this PR allows the detection of smaller page flings. Is that right?
 
I couldn't reproduce the pattern with small fling. 
When We do a big swipe, the page value sometimes overflows from the target index, and I think that's the cause of this problem.
 
page overflows with big fling |
:-: | 
 |

Comment (justinmc) [reply packages/flutter/test/material/tabs_test.dart:644]: Got it, thank you for the thorough response and the diagram!  I was missing the fact that this fling was more than one whole item in length.  Makes sense.  I also confirmed that this test fails on master (it gives `0` on master, as expected from your diagram).

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Can you merge master into this PR?  You have some check failures, but I think they are unrelated and will hopefully be fixed after you merge in master.

**Comment by HeavenOSK** [comment]
@justinmc 
Thank you very much for your advice, I learned a lot from your comments! It was a good chance for me to write UI Test.
 
I merged `mater` of upstream into the branch `fix-tabview-decrement`. https://github.com/flutter/flutter/pull/88878/commits/1f56c4f150bc0862a7ed92b70a57d84a0a43e0b9

**Comment by justinmc** [comment]
No problem!  Could you merge master in again, actually?  Sorry, I think something is wrong with the checks.  A bunch of them seem to be stuck at "Queued"...

**Comment by justinmc** [comment]
That did it, thank you!  I will merge this after the Google tests finish.  I don't expect a problem there.

**Comment by HeavenOSK** [comment]
@justinmc 
Thank you so much!

**Comment by justinmc** [comment]
No problem!  Looks like the Google testing status isn't updating for some reason, I'll merge the PR.

**Pull request merged**
